File: clojure/day3/problem2.clj
(ns user
  (:require [clojure.string :as str]))

(defn input []
  (->> "input.txt"
       slurp
       str/split-lines
       (map str/trim)))

(def test-input
  (->> "987654321111111
811111111111119
234234234234278
818181911112111"
       str/split-lines
       (map str/trim)))

(defn parse-bank [line]
  (mapv #(Integer/parseInt (str %)) line))

(defn solve [acc nb bank]
  (if (zero? nb)
    acc
    (let [split-index     (- (count bank) (dec nb))
          candidates      (take split-index bank)
          [next-idx max]  (->> (map-indexed vector candidates)
                               (sort-by second #(compare %2 %1))
                               first)
          remaining-banks (drop (inc next-idx) bank)]
      (recur (conj acc max) (dec nb) remaining-banks))))

(defn calc [xs]
  (->> xs
       (mapv parse-bank)
       (mapv #(solve [] 12 %))
       (map (fn [digits] (Long/parseLong (apply str digits))))
       (reduce +)))


(calc test-input)
;; => 3121910778619

(calc (input))
;; => 171741365473332

File: fsharp/day3/problem2.fs
open System.IO

let input () = 
    let path = Path.Combine(Directory.GetCurrentDirectory(), "input.txt")
    File.ReadAllLines(path)
    |> Array.map (fun s -> s.Trim())
    |> List.ofSeq

let testInput = 
    """987654321111111
811111111111119
234234234234278
818181911112111""".Split("\n")
    |> Array.map (fun s -> s.Trim())
    |> List.ofSeq

let parseBank (line: string) =
    line
    |> Seq.map (string >> int)
    |> Seq.toList

let rec solve acc nb (bank : int list) =
    if nb = 0 then acc |> List.rev
    elif nb = bank.Length then (acc |> List.rev) @ bank
    else 
        let splitIndex = bank.Length - (nb - 1)
        let candidates = List.take splitIndex bank
        let (idxnext, next) = candidates |> List.indexed |> List.maxBy snd
        let remainingBank = bank |> List.skip (idxnext + 1)
        solve (next :: acc) (nb - 1 ) remainingBank


let solver xs =
    xs
    |> List.map parseBank
    |> List.map (solve [] 12)
    |> List.map (List.map string >> String.concat "" >> int64)
    |> List.sum

testInput
|> solver

input ()
|> solver
// 171741365473332 is the correct value